File: Neetcode 150/Arrays and Hashing/valid-sudoku.cpp

// https://leetcode.com/problems/valid-sudoku/description/

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        int n = board.size();
        unordered_set<int> rowSet[n], colSet[n], boxSet[n];

        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                auto val = board[i][j];
                if(val == '.') continue;

                int boxNumber = 3*(i/3) + (j/3);

                if(rowSet[i].find(val) != rowSet[i].end()) return false;
                if(colSet[j].find(val) != colSet[j].end()) return false;
                if(boxSet[boxNumber].find(val) != boxSet[boxNumber].end()) return false;

                rowSet[i].insert(val);
                colSet[j].insert(val);
                boxSet[boxNumber].insert(val);
            }
        }

        return true;   
    }
};

File: Neetcode-150-js/Arrays and Hashing/valid-sudoku.js

// https://leetcode.com/problems/valid-sudoku/description/

/**
 * @param {character[][]} board
 * @return {boolean}
 */

const LEN = 9;

var isValidSudoku = function(board) {
    
    const rowSet = Array.from({length : LEN}, () => new Set());
    const colSet = Array.from({length : LEN}, () => new Set());
    const boxSet = Array.from({length : LEN}, () => new Set());

    for(let i = 0; i < LEN; i++){
        for(let j = 0; j < LEN; j++){
            let val = board[i][j];
            if(val === ".") continue;
            let boxNumber = 3*Math.floor(i/3) + Math.floor(j/3);

            if(rowSet[i].has(val) || colSet[j].has(val) || boxSet[boxNumber].has(val)) return false;

            rowSet[i].add(val);
            colSet[j].add(val);
            boxSet[boxNumber].add(val);
        }
    }

    return true;
};